galley/pkg/server/components/status.go

```diff
@@ -0,0 +1,382 @@
+				return client.ExtensionsV1beta1().Ingresses(args.WatchedNamespace).List(opts)
+			},
+			WatchFunc: func(opts metaV1.ListOptions) (watch.Interface, error) {
+				return client.ExtensionsV1beta1().Ingresses(args.WatchedNamespace).Watch(opts)
+			},
+		},
+		&v1beta1.Ingress{}, args.ResyncPeriod, cache.Indexers{},
+	)
+
+	st := StatusSyncer{
+		client:         client,
+		informer:       informer,
+		queue:          queue,
+		mode:           mode,
+		ingressClass:   ingressClass,
+		ingressService: meshConfig.IngressService,
+		handler:        handler,
+	}
+
+	st.elector = st.buildElector(st, electionID)
+
+	// Register handler at the beginning
+	ingressNamespace := constants.IstioIngressNamespace
+	handler.Append(func(obj interface{}, event model.Event) error {
+		addrs, err := st.runningAddresses(ingressNamespace)
+		if err != nil {
+			return err
+		}
+
+		return st.updateStatus(sliceToStatus(addrs))
+	})
+
+	return &st
+}
+
+func (s *StatusSyncer) Start() (err error) {
+	s.informerStopCh = make(chan struct{})
+	go s.informer.Run(s.informerStopCh)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	s.electorCancel = cancel
+	go s.elector.Run(ctx)
+
+	return nil
+}
+
+func (s *StatusSyncer) Stop() {
+	if s.informerStopCh != nil {
+		close(s.informerStopCh)
+		s.informerStopCh = nil
+	}
+
+	if s.electorCancel != nil {
+		s.electorCancel()
+		s.electorCancel = nil
+	}
+}
+
+func (s *StatusSyncer) buildElector(st StatusSyncer, electionID string) *leaderelection.LeaderElector {
+	callbacks := leaderelection.LeaderCallbacks{
+		OnStartedLeading: func(ctx context.Context) {
+			log.Infof("I am the new status update leader")
+			go st.queue.Run(ctx.Done())
+			err := wait.PollUntil(updateInterval, func() (bool, error) {
+				st.queue.Push(kube.NewTask(st.handler.Apply, "Start leading", model.EventUpdate))
+				return false, nil
+			}, ctx.Done())
+
+			if err != nil {
+				log.Errorf("Stop requested")
+			}
+		},
+		OnStoppedLeading: func() {
+			log.Infof("I am not status update leader anymore")
+		},
+		OnNewLeader: func(identity string) {
+			log.Infof("New leader elected: %v", identity)
+		},
+	}
+
+	broadcaster := record.NewBroadcaster()
+	hostname, _ := os.Hostname()
+
+	recorder := broadcaster.NewRecorder(scheme.Scheme, coreV1.EventSource{
+		Component: "ingress-leader-elector",
+		Host:      hostname,
+	})
+
+	lock := resourcelock.ConfigMapLock{
+		ConfigMapMeta: metaV1.ObjectMeta{Namespace: "default", Name: electionID},
+		Client:        st.client.CoreV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity:      podNameVar.Get(),
+			EventRecorder: recorder,
+		},
+	}
+
+	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          &lock,
+		LeaseDuration: electorTTL,
+		RenewDeadline: electorTTL / 2,
+		RetryPeriod:   electorTTL / 4,
+		Callbacks:     callbacks,
+	})
+
+	if err != nil {
+		log.Errorf("unexpected error starting leader election: %v", err)
+	}
```

Comment:
I think we should pass err up the chain instead of eating it. This seems like it should be another case where `NewStatusSyncer` should end up returning nil.

Reply:
sounds good!